smarts/env/wrappers/frame_stack.py

```diff
@@ -0,0 +1,77 @@
+# MIT License
+#
+# Copyright (C) 2021. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import copy
+import gym
+from collections import deque
+from typing import Deque, Dict, Tuple, Union
+from smarts.core import sensors
+
+
+class FrameStack(gym.Wrapper):
+    """ By default, this wrapper will stack 3 consecutive frames as an agent observation"""
+
+    def __init__(self, env: gym.Env, num_stack: int = 3):
+        assert num_stack > 1, f"Expected num_stack > 1, but got {num_stack}."
+        super(FrameStack, self).__init__(env)
+        self.num_stack = num_stack
+        self.frames = {
+            key: deque(maxlen=self.num_stack) for key in self.env.agent_specs.keys()
+        }
+
+    def _get_observations(
+        self, frame: sensors.Observation
+    ) -> Dict[str, Deque[sensors.Observation]]:
+        """Update and return frames stack with given latest single frame."""
+
+        new_frames = dict.fromkeys(frame)
+
+        for agent_id, observation in frame.items():
+            self.frames[agent_id].append(observation)
+            frames_list = list(self.frames[agent_id])
+            new_frames[agent_id] = copy.deepcopy(frames_list)
```

Comment:
I understand why you're doing it, but this deep copy is sort of expensive (especially if `num_stack` is large).  I would recommend adding a comment to the class explaining this (that is, explaining that the stack will incur the cost of deep copying each frame `num_stack` times).

While you're at it, it should also be explained that the stack is a "moving window" on the frames... that is, the same frame will be in more than one result, just shifted over by one spot each time.